### PR TITLE
Add basic inputs generator

### PR DESCRIPTION
**Pull request by davidleonfdez**
It basically merges all the poems into two text files, one for the training set and another for the validation set.

This done by _notebooks/txt_inputs_generator.ipynb_. Apart from the txts, it generates an additional file that stores the formatting options chosen by the user. That information should be used later by the tokenizer, for example, to add special tokens, but this way the generation of input files is decoupled from the training process.

Of course, the main limitation we have is the GPT2 max sequence length of 1024 tokens, so the number of extra tokens included in every sequence should be limited. In fact, the option _VerseGrouping.OnePoemBySequence_ may not make much sense, as it exacerbates the problem and will lead to the truncation of most poems.

The "n_previous_verse_terminations" param is an attempt to help the network **learn rhymes** but I'm not sure if it's the best way to make it work in practice. 
If n_previous_verse_terminations = "a", it means that, before the beginning of verse "i", in the same line, the last word of the previous "a" verses must be included 

> ([last token of verse i-a], [last token of verse i-a+1], ..., [last token of verse i-1], <beginning_of_verse_token>, [tokens of verse i].

(You can see an example in tests/data_tests.py too)

It's theoretically possible that it works in combination with some form of conditioning based on poem form/structure, given that those inserted words will be affected in order by the positional embedding, but I'm not sure it's easy for the NN to learn it.

While training, maybe it could improve if only the last token of each verse is allowed to attend to those terminations, but I don't think it's feasible with HuggingFace (at least, without overwriting GPT2 code). The GPT2 input parameter 'attention_mask' is actually a "padding mask" (batch_size * sequence_length), but we'd need to change the mask of size (sequence_length * sequence_length) that indicates which positions of the sequence are allowed to attend to which positions (by default, for causal LMs like this, each token is allowed to attend to all the previous tokens of its sequence). I guess the mask we look for is the "bias" buffer of GPT2Attention 
(see https://github.com/huggingface/transformers/blob/6ac77534bfe97c00e0127bb4fc846ae0faf1c9c5/src/transformers/models/gpt2/modeling_gpt2.py ), which seems to be kind of hardcoded to be the same for every sequence.

Anyway, despite not being able to modify the causal mask, we could still look at the output attention weights of same example sentences, in order to check if only the last token of a verse attends to the terminations of the previous verses. Even if it works, I expect at least some marginal attention from other tokens because, apart from the rhymes, of course some meaning can still be extracted from verse endings. 

A bit offtopic, but now I'm realizing attention weights are indeed something that could be useful for the visualizations, (if you are not using them already) @ismaelfaro.

### CODE REVIEW

**Comment by davidleonfdez** [comment]
The more I think about it, the more I dislike my idea of including the last token of previous verses before every verse, because anyway the preprocessor of the happytransformer trainer concatenates lines to form a sequence until reaching the maximum sequence length, so in many cases the information is redundant. 

It's only useful for initial verses of a sequence or extremely long verses.

Anyway, I'll test it.

**Comment by ismaelfaro** [comment]
cool, everything make sense.

Something that I'll add would be part of this details in the readme.md, at least the summary, and if you thing that can be interesting you can create a doc folder where collect all the details to use the Notebooks for that. 

Also your comment about the HuggingFace GTP2 transformer could be a good point of collaboration. 

and the idea of use the output attention weights is so cool, we should play with that for sure.

**Comment by davidleonfdez** [comment]
Now a csv with the details of the assignment of poems to splits is also generated. We could need to know this, for example, to train a classifier with the same splits.
